### ISSUE TITLE
Book solution of FLAT cannot be admitted

### ISSUE THREAD

**Issue by rongcuid** [acl2/acl2#1596]
Not sure if this is the right place for an issue, but the ACL2 book's solution for flat (pg. 108 An Approach) doesn't admit in ACL2 8.5:

```lisp
(defun flat (x)
  (declare (xargs :measure
                  (cons (1+ (acl2-count x))
                        (acl2-count (car x)))))
  (cond ((atom x) (list x))
        ((atom (car x)) (cons (car x) (flat (cdr x))))
        (t (flat (cons (caar x)
                       (cons (cdar x) (cdr x)))))))
For the admission of FLAT we will use the relation O< (which is known
to be well-founded on the domain recognized by O-P) and the measure
(CONS (+ 1 (ACL2-COUNT X)) (ACL2-COUNT (CAR X))).  The non-trivial
part of the measure conjecture is

Goal
(AND (O-P (CONS (+ 1 (ACL2-COUNT X))
                (ACL2-COUNT (CAR X))))
     (IMPLIES (AND (NOT (ATOM X))
                   (NOT (ATOM (CAR X))))
              (O< (CONS (+ 1
                           (ACL2-COUNT (LIST* (CAR (CAR X))
                                              (CDR (CAR X))
                                              (CDR X))))
                        (ACL2-COUNT (CAR (LIST* (CAR (CAR X))
                                                (CDR (CAR X))
                                                (CDR X)))))
                  (CONS (+ 1 (ACL2-COUNT X))
                        (ACL2-COUNT (CAR X)))))
     (IMPLIES (AND (NOT (ATOM X)) (ATOM (CAR X)))
              (O< (CONS (+ 1 (ACL2-COUNT (CDR X)))
                        (ACL2-COUNT (CADR X)))
                  (CONS (+ 1 (ACL2-COUNT X))
                        (ACL2-COUNT (CAR X)))))).
Subgoal 3
Subgoal 3'

([ A key checkpoint:

Subgoal 3'
NIL

A goal of NIL, Subgoal 3', has been generated!  Obviously, the proof
attempt has failed.

])


ACL2 Error in ( DEFUN FLAT ...):  The proof of the measure conjecture
for FLAT has failed.



Summary
Form:  ( DEFUN FLAT ...)
Rules: ((:DEFINITION ATOM)
        (:DEFINITION NOT)
        (:DEFINITION O-FINP)
        (:DEFINITION O-P)
        (:FAKE-RUNE-FOR-TYPE-SET NIL)
        (:REWRITE CAR-CONS)
        (:TYPE-PRESCRIPTION ACL2-COUNT))
Time:  0.00 seconds (prove: 0.00, print: 0.00, other: 0.00)
Prover steps counted:  322

---
The key checkpoint goal, below, may help you to debug this failure.
See :DOC failure and see :DOC set-checkpoint-summary-limit.
---

*** Key checkpoint at the top level
    before generating a goal of NIL (see :DOC nil-goal): ***

Subgoal 3'
NIL

ACL2 Error [Failure] in ( DEFUN FLAT ...):  The proof of the measure
conjecture for FLAT has failed.  See :DOC failure.

******** FAILED ********
```

**Comment by MattKaufmann** [comment]
Hi,

Thank you for bringing that error to our attention.  It's because the
notion of ACL2 ordinal changed several years ago, after publication of
the book.  Here is a corrected version.

(defun flat (x)
  (declare (xargs :measure
                  (list (1+ (acl2-count x))
                        (acl2-count (car x)))
                  :well-founded-relation l<))
  (cond ((atom x) (list x))
    ((atom (car x)) (cons (car x) (flat (cdr x))))
    (t (flat (cons (caar x)
               (cons (cdar x) (cdr x)))))))

I'll get with my co-authors about what sort of Errata to put out.

By the way, although the list you sent to works, probably the
acl2-help list would be better for queries like this one, that is,
about ACL2 usage (including mis-use, as in the book!).  You can join
acl2-help via a link on the ACL2 home page
(http://www.cs.utexas.edu/users/moore/acl2/); let me know if you run
into any problems doing so.

Thanks,
Matt

**Comment by MattKaufmann** [comment]
Hi Rongcui,

Thank you for reporting the issue for the flat function on page 108 of
the "An Approach" book.  J pointed out to me that the errata page for
the two ACL2 print books

  https://www.cs.utexas.edu/~moore/publications/acl2-books/errata.html

actually mentioned the following, which hints at a solution.

  For example, the representation of ordinals has changed and to revert
  ACL2 to the representation used 20 years ago one must include these
  two events early in the session.

  (include-book "ordinals/e0-ordinal" :dir :system)
  (set-well-founded-relation e0-ord-<)

To add clarity, we have added the following sentence to the end of
that page (again, thank you for pointing out the issue for the flat
function).

  For example, the definition of flat in Section 7.3 of Computer-Aided
  Reasoning: An Approach can then be admitted to ACL2.

We have also added correspondingly to other errata pages for the two
specific books.

-- Matt
